compiler/rustc_mir_build/src/thir/pattern/mod.rs

```diff
@@ -237,6 +244,12 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
                 };
             }
         }
+        for inline_const in [lo_inline, hi_inline] {
+            if let Some(def) = inline_const {
+                kind =
+                    PatKind::InlineConstant { def, subpattern: Box::new(Pat { span, ty, kind }) };
```

Comment:
I don't think it's that intuitive to hide the range pattern behind this nested `PatKind::InlineConstant` construction. Having an explicit range pattern is more explicit imo. How are inline constants in range patterns currently handled in later stages? Does this change just move the complexity from a later stage to THIR?

Reply:
This makes things simpler for the consuming side: unsafety checking only cares that there are some inline constants around somewhere, pattern checking only cares that there's a range pattern whose extremities can be evaluated to bits. Carrying the unevaluated constant around felt invisible and brittle, my gut said danger.

But I am biased. I have an open PR (https://github.com/rust-lang/rust/pull/116692) that reworks `PatRange` a lot, and I'm hoping I can eventually evaluate the consts eagerly and store bits in `PatRange`.

Reply:
> Carrying the unevaluated constant around felt invisible and brittle, my gut said danger.

I don't understand this. Looking at `const_to_pat`, we always error if we find an unevaluatable constant. When we lower inline consts we currently always go through `const_to_pat` in the non-error case. Why would we carry unevaluated constants around?

Can't we introduce a new enum, that is used in `Patkind::Constant` and includes variants for `InlineConst(mir::Const, LocalDefId)` and `mir::Const` instead of modelling inline constants on the pattern level? The visitor in thir unsafety checking could still detect inline constants, pattern checking should be able to work with this afaict and we could continue to encode pattern ranges with inline constants as `PatKind::Range` directly, which seems less hacky than the nested PatKind::InlineConst imo.

Reply:
As I said above
> Inline constants, like other constants, can be lowered to actual patterns in THIR. We could keep them as constants here and lower to patterns in both MIR building and exhaustiveness checking, but I don't think that's any better.

Reply:
> Why would we carry unevaluated constants around?

Unsafety checking needs to look at the body of the inline constant, so if we want to use `Const` to carry it around, we must not evaluate it. Pretty sure an evaluated const doesn't know about the body it came from. This is what was done in ranges before I asked for this special `InlineConstant` behavior.

> Can't we introduce a new enum, that is used in Patkind::Constant

We can't if we have an inline constant that evaluates to something that `const_to_pat` will deconstruct, e.g.:
```rust
fn main() {
    match Some(0) {
        const { Some(0) } => println!("got 0!"),
        Some(0) => unreachable!(),
        _ => println!("something else"),
    }
}
```
Here the constant needs to be deconstructed into a `Pat` that represents `Some(0)` so that we can do proper exhaustiveness/unreachable checking with it. The resulting `Pat` might not have any `PatKind::Constant` left at the end. Hence why we attach the inline constant in the middle of the pattern.